src/constraint_system/composer.rs

```diff
@@ -585,20 +587,23 @@ impl TurboComposer {
         let b = self.witnesses[&b];
         let d = self.witnesses[&d];
 
-        let qm = s.coeff(Selector::Multiplication);
-        let ql = s.coeff(Selector::Left);
-        let qr = s.coeff(Selector::Right);
-        let qd = s.coeff(Selector::Fourth);
-        let qc = s.coeff(Selector::Constant);
+        let q_m = s.coeff(Selector::Multiplication);
+        let q_l = s.coeff(Selector::Left);
+        let q_r = s.coeff(Selector::Right);
+        let q_d = s.coeff(Selector::Fourth);
+        let q_c = s.coeff(Selector::Constant);
         let pi = s.coeff(Selector::PublicInput);
 
-        let x = qm * a * b + ql * a + qr * b + qd * d + qc + pi;
-
-        let y = s.coeff(Selector::Output);
-        let y: Option<BlsScalar> = y.invert().into();
-        let y = -y.expect("Inconsistent internal usage of `Constraint::evaluate_output`: Output selector should not be zero");
+        let dividend = q_m * a * b + q_l * a + q_r * b + q_d * d + q_c + pi;
+        let q_o: BlsScalar = *s.coeff(Selector::Output);
+        let divisor_inverse = if q_o * q_o == BlsScalar::one() {
```

Comment:
Something sounds off. We perform an extra scalar multiplication for no clear reason when we could check individually for either `-1` or `1`.

Also, we have a clear tech debt in that `expect`: a malicious user can crash a plonk server by submitting a wrong proof. If we are to refactor this part, we should either change the signature of the function to `pub(crate) fn append_output_witness(&mut self, s: Constraint) -> Witness`, or set a constant when `q_o` happens to be zero.

Changing the signature is preferable because it will reflect what happens internally: any proof request with `q_o == 0` is invalid

Reply:
I agree with the that the extra scalar multiplication is not needed here, will check for `1` and `-1` individually.

Not sure what you mean with the change of the signature? I assume you mean that we return a `Result<Witness, Error>` ?
I would also avoid setting q_o to `0` manually since it should not be `0` at this point and if it is, it might point to an error somewhere else.

Reply:
Thinking of it, why do we actually not check for `1` and `-1` directly in `invert()`? Might give us some performance increase at other spots too

Reply:
Its quite easy to have `q_o` as `0` here. The user just have to not set anything in the output wire.

Reply:
This moved to [bls12_381](https://github.com/dusk-network/bls12_381/issues/78)